dev/io.openliberty.security.authorization.internal.jacc.3.0/src/io/openliberty/security/authorization/jacc/internal/proxy/ProviderServiceProxyImpl.java

```diff
@@ -85,50 +112,53 @@ public String getFactoryName() {
     }
 
     protected void activate(ComponentContext cc) {
+        locationAdminRef.activate(cc);
         jaccProviderService.activate(cc);
     }
 
     protected void deactivate(ComponentContext cc) {
+        // unset PolicyFactory so we do not leak the user feature classloader for the ProviderService
+        PolicyFactory.setPolicyFactory(null);
```

Comment:
Would love to learn more about that.

Reply:
If the bundle that has the PolicyFactory is not stopped, it will keep a reference to the Policy objects from the user feature so if the user feature is stopped, we will end up leaking the ClassLoader from the user feature.  This was brought up by Tom Watson after the UFO socialization and is in the list of things to consider from the socialization.  It is even worse in older versions of Jakarta Authorization / JACC because the Policy class is in the JDK and that won't be stopped at all.  Moreover when the user application can set the Policy / PolicyFactory object in the API when the application is stopped, the PolicyFactory class can hold a reference to the application classloader that was stopped.  So this is a fail safe to make sure that we don't leak these classloaders.